refactor: fixes to MVVM Architecture followed by the app

adityastic opened the pull request:
Fixes #178 

Changes:
- ViewModels for Course and Video Activity
- Updated all Null Assertions with Safe Calls
- Improved Resources

adityastic commented:
@aggarwalpulkit596 have a look

adityastic commented:
@aggarwalpulkit596 The Online API Services are not working. The Travis Failed, Please Check if the API's are up or not

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/activities/CourseActivity.kt:
```diff
@@ -204,85 +212,100 @@ class CourseActivity : AppCompatActivity(), AnkoLogger {
-                        addtocart(id, name)
+        viewModel.getCourse(object : OnCompleteCourseListener {
+            @Suppress("UNCHECKED_CAST")
+            override fun onCompleteJson(response: Response<JsonObject>) {
```
Don't use listeners/callback method use livedata instead see other viewmodels for reference

adityastic replied:
on it

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/activities/CourseActivity.kt:
```diff
@@ -204,85 +212,100 @@ class CourseActivity : AppCompatActivity(), AnkoLogger {
-                                        rvExpandableskeleton.hide()
-                                        sectionAdapter.setData(sectionsList)
+                    fetchTags(course)
+                    showPromoVideo(course.promoVideo)
```
Even this should be directly fetched from livedata value see MyCourseActitvity

adityastic replied:
let me push an update to this as well

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt:
```diff
@@ -0,0 +1,74 @@
+}
+
+interface OnVideoCompleteListener {
+    fun onComplete(response: Response<JsonObject>)
```
And why JsonObject obviously it won't work

adityastic replied:
removed

adityastic replied:
im updating the course activity, give your feedback on the changes. If it gets accepted I shall go on and perform the same on the Video Activity

aggarwalpulkit596 replied:
First of all they are Gson Response not Json Response and secondly it shouldn't be like that it should be at all you should save the dao here itself and if error show an error message using a boolean livedata variable and snackbar message

adityastic replied:
I didn't change the Video Activity as of now. I've updated only the Course Activity and it's ViewModel with the livedata and observers with the mutable boolean variables. I will make changes to the VideoClass once you confirm the refactor for One Class

adityastic replied:
Btw I don't get what you mean by Gson Response. It is a Retrofit Response of Gson JsonObject right?

aggarwalpulkit596 replied:
No it's GSON see client class in onlineapi. Module

adityastic replied:
@aggarwalpulkit596 Check this Sir.

The Call
<img width="348" alt="Screenshot 2019-05-30 at 2 37 02 PM" src="https://user-images.githubusercontent.com/11988517/58622140-8fe9d280-82e8-11e9-8f03-45d79dd0a7ff.png">

The GSON JsonObject import
<img width="360" alt="Screenshot 2019-05-30 at 2 37 10 PM" src="https://user-images.githubusercontent.com/11988517/58622137-8d877880-82e8-11e9-9b5d-3c0fc6de36ec.png">

aggarwalpulkit596 replied:
Not every call is a JsonObject notes are of type GSON

aggarwalpulkit596 replied:
There are two different interfaces onlineRESTAPI and onlineJSONAPI so do your work accordingly

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt:
```diff
@@ -0,0 +1,74 @@
+    var runDao: CourseRunDao
+) : ViewModel() {
+    lateinit var attemptId: String
+    lateinit var contentId: String
```
Move these variables to Activity

adityastic replied:
done

adityastic commented:
@aggarwalpulkit596 updated CourseActivity, confirm if the changes are as per the requirement.

adityastic commented:
@aggarwalpulkit596 stacktrace to the failure:
`com.codingblocks.onlineapi.api.OnlineJsonApiPublicTest > GET recommended FAILED
    java.lang.AssertionError at OnlineJsonApiPublicTest.kt:30`

Please Confirm the issue with the public API

aggarwalpulkit596 commented:
> @aggarwalpulkit596 stacktrace to the failure:
> `com.codingblocks.onlineapi.api.OnlineJsonApiPublicTest > GET recommended FAILED java.lang.AssertionError at OnlineJsonApiPublicTest.kt:30`
> 
> Please Confirm the issue with the public API

Check Now

adityastic commented:
@aggarwalpulkit596 fixed, check the CourseActivity. Let me know if you need any more changes in the approach

adityastic commented:
@aggarwalpulkit596 I've updated both the activities. Thanks for the update

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+                } catch (e: Exception) {
+                    getOtpProgress.value = false
+                }
+                getOtpProgress.value = (response?.isSuccessful == true)
```
otpProgress ?

adityastic replied:
Forgot to remove it after setting observer, thanks

adityastic replied:
Yes when response is received, we need to initialize the view. So it acts as a listener as you said

aggarwalpulkit596 replied:
how are notes related to otpProgress?

adityastic replied:
Oh wait, my bad. I intialized the wrong variable. Updated, have a look

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+                try {
+                    notesDao.insert(
+                        NotesModel(
+                            response?.body()?.id ?: "",
```
`use response?.body?.let {}`

adityastic replied:
Got it

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+    fun createNote(note: Notes) {
+        Clients.onlineV2JsonApi.createNote(note)
+            .enqueue(retrofitCallback { _, response ->
+                try {
```
why only notes are in try catch block and not others ?

adityastic replied:
It is because the original file had a try block in notes, and others didn't

adityastic replied:
I've updated the rest dao operations as well :)

aggarwalpulkit596 replied:
i made some changes in this pr why had you removed them?

adityastic replied:
@aggarwalpulkit596 Must have occured because I force pushed the branch, can you suggest the changes again?

aggarwalpulkit596 replied:
you're inserting notes and doubt irrespective of the response is successful or not it should be something like `if(response.isSuccesful)`

adityastic commented:
@aggarwalpulkit596 updated, have a look if I am missing anything

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt:
```diff
@@ -0,0 +1,82 @@
+                    if (response.isSuccessful) {
+                        doubtsDao.insert(
+                            DoubtsModel(
+                                doubt.id, doubt.title, doubt.body, doubt.contents?.id
```
it should be it.id as doubt.id is not auto-generated and is returned from the server.
Check the complete flow of the app make sure there is no crash/bug etc.

adityastic replied:
Yes as I'm using let, I should use it here. Let me test the complete app again to verify everything is working. Give me an hour to make sure it is not glitched

aggarwalpulkit596 replied:
No, it's not about using let here doubt.id would be null as while creating an object of doubt we don't have an id it is generated from server side and then we save it do our DB

aggarwalpulkit596 replied:
same goes for notes as well

aggarwalpulkit596 replied:
also refractor VideoDoubtsFragment and VideoNotesFragment and delete the singleton object of appdatabase and make sure appdatabase object is being no where and only koin is being used for accessing database
Don't worry I'll accommodate points accordingly

adityastic replied:
@aggrawalpulkit596, sounds great. This pr was very informative btw, thanks for having me here

adityastic replied:
@aggarwalpulkit596 tell me a course which has doubts or notes in it

aggarwalpulkit596 replied:
Notes and doubts are user created thing not a course specific thing

adityastic replied:
@aggarwalpulkit596 yes please name one course for me, I need to test them

aggarwalpulkit596 replied:
you can enroll in any course for trial and check the functionality

adityastic replied:
@aggarwalpulkit596 great, let me try that

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/adapters/CourseDataAdapter.kt:
```diff
@@ -48,51 +48,42 @@ class CourseDataAdapter(
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
 
-        var viewHolder: RecyclerView.ViewHolder? = null
+        var viewHolder: RecyclerView.ViewHolder = MyCoursesViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.my_course_card_horizontal, parent, false))
```
Why . are you setting viewHolder even before knowing the type

adityastic replied:
when you know there are only two conditions, set the initial value as one and change it if it matches the other.

adityastic replied:
it is better than null asserting the viewholder which is a crucial component

adityastic replied:
I've improved the condition, let me know if you agree

adityastic commented:
@aggarwalpulkit596 wow, it was a fun day after all. Here are the changes :
- Created ViewModels for 6 Classes
- Removed Null Calls from all the changed classes
- Removed AppDatabase Instance 
- fixed explicit scoping for all classes
etc
Please have a look if I missed anything

adityastic commented:
@aggarwalpulkit596 updates on this plz

aggarwalpulkit596 commented:
Will take a look tomorrow till then you can do can open some more issues or work on other pending issues

adityastic commented:
@aggarwalpulkit596 morning, sure take your time. I shall work on the other issues till then. Have a nice day

adityastic commented:
@aggarwalpulkit596 please review

aggarwalpulkit596 commented:
Will update in an hour

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/MyCourseViewModel.kt:
```diff
@@ -22,8 +22,8 @@ import com.codingblocks.onlineapi.Clients
     var runDao: CourseRunDao,
-    private var sectionWithContentsDao: SectionWithContentsDao,
-    private var contentsDao: ContentDao,
+    var sectionWithContentsDao: SectionWithContentsDao,
```
why remove private from them

adityastic replied:
because I did not take the mutable approach. I'm updating it as well as all the viewmodels to follow this

adityastic replied:
because I passed them as parameters into the adapter, let me find the perfect way out

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/services/DownloadService.kt:
```diff
@@ -57,15 +58,13 @@ class DownloadService : IntentService("Download Service"), AnkoLogger,
         notificationBuilder.setContentTitle(title)
-        database = AppDatabase.getInstance(this)
-        contentDao = database.contentDao()
+        contentDao = AppDatabase.getInstance(this).contentDao()
```
there should be not dependacy of AppDatabase anywhere
Infact i told you to delete the singleton object and use koin everywhere

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/fragments/VideoNotesFragment.kt:
```diff
@@ -57,77 +48,29 @@ class VideoNotesFragment : Fragment(), AnkoLogger {
         view.notesRv.addItemDecoration(itemDecorator)
 
-        notesDao.getNotes(param1!!).observer(this) {
+        viewModel.notesDao.getNotes(param1.toString()).observer(this) {
```
it should access it livedata variable and not notesDao 
If there are viewModels for fragment/activity use livedata or mutables variables dao  shouldn't be accessible to fragments/activites it should be a private member of viewmodel

adityastic replied:
Got it

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/fragments/VideoDoubtFragment.kt:
```diff
@@ -42,45 +33,23 @@ class VideoDoubtFragment : Fragment(), AnkoLogger {
         view.doubtsRv.addItemDecoration(itemDecorator)
-        doubtsDao.getDoubts(param1!!).observe(this, Observer<List<DoubtsModel>> {
+
+        viewModel.doubtsDao.getDoubts(param1.toString()).observe(this, Observer<List<DoubtsModel>> {
```
use Observer extension and use live data variable not dao function

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/fragments/MyCoursesFragment.kt:
```diff
@@ -51,7 +51,8 @@ class MyCoursesFragment : Fragment(), AnkoLogger {
             CourseDataAdapter(
                 ArrayList(),
-                activity!!,
+                viewModel.courseDao,
```
why this ?

adityastic replied:
You were passing the activity, that is never a good practise. Check the line below this, it is the best practise to use context in any fragment

adityastic replied:
And the dao. Because we were initializing the dao inside the adapter everytime a view was binding.

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/fragments/AnnouncementsFragment.kt:
```diff
@@ -9,25 +9,27 @@ import androidx.lifecycle.Observer
         super.onCreate(savedInstanceState)
         arguments?.let {
-            courseId = it.getString(ARG_COURSE_ID)!!
+            courseId = it.getString(ARG_COURSE_ID) ?: ""
```
use variable from constant class

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/adapters/VideosDoubtsAdapter.kt:
```diff
@@ -168,10 +164,8 @@ class VideosDoubtsAdapter(private var doubtsData: ArrayList<DoubtsModel>) :
-                        notifyDataSetChanged()
-                    }
+                .enqueue(retrofitCallback { _, _ ->
+                    notifyDataSetChanged()
```
notify only if the response is successful

aggarwalpulkit596 commented on app/src/main/java/com/codingblocks/cbonlineapp/adapters/SectionDetailsAdapter.kt:
```diff
@@ -420,11 +414,11 @@ class SectionDetailsAdapter(
         }
 
-        private fun connectedToWifi(context: Context): Boolean {
+        private fun connectedToWifi(context: Context): Boolean? {
```
move this to utils

adityastic commented:
@aggarwalpulkit596 I've Updated all the ViewModels. Please allot me bounty accordingly. Thanks

aggarwalpulkit596 commented:
Notes are not working

aggarwalpulkit596 commented:
Check Each and Every Activity/Fragment and add a video for the same

adityastic commented:
@aggarwalpulkit596 is the doubts api not working?

adityastic commented:
@aggarwalpulkit596 https://drive.google.com/file/d/1-74XapXg_Fs3bVFdX10LovOtqvVqgXKJ/view?usp=drivesdk

adityastic commented:
@aggarwalpulkit596 please have a look

adityastic commented:
@aggarwalpulkit596 I've updated the PR, changed the null assertions in shortcuts.

aggarwalpulkit596 commented:
Why are you making changes which are not required in this PR i wouldn't be giving you any extra points for this and will not accept any PR in future if any unnecessary changes are made

adityastic commented:
@aggarwalpulkit596 It was causing merge conflicts as the shorcuts are created in HomeActivity and I made changes to it

adityastic commented:
@aggarwalpulkit596 I can understand this PR is so big already, If you want I can seperate this into multiple PR's so we can target one part of the application at a time and it would release the load from you as well. Let me know your thoughts on this

aggarwalpulkit596 commented:
No issues i will merge this one today positively but for future don't do changes which are not related to this pr i specifically told you not convert all the classes to viewmodels and still you did that no issues it's a good thing but a author it's makes my work more and more difficult to check the complete PR probably i will miss few things in this PR which i have to fix myself

adityastic commented:
@aggarwalpulkit596 You can always find me here if there is any issue hindering this pr. I shall be available whenever you need me

aggarwalpulkit596 commented:
In Some Adapter, you have passed viewmodel object where as in some adapters you're passing dao object use one of them only and also instead of toString use elvis operator  `? : ""`

aggarwalpulkit596 commented:
Rest Looks Good To Me Although there are some irrelevant changes in the PR 1500 points would be sufficient i guess
Points are not proportional to amount to work as there 1000+ lines but it is actually the amount of actual productive work/ adding a new feature fixing a bug etc so 1500 would be fine

adityastic commented:
@aggarwalpulkit596 Yes I completely agree, productivity is never proportional to LOC, I appreciate the extra 500. Thank you so much 😆

adityastic commented:
@aggarwalpulkit596 regarding the adapter with dao object. It is specifically with the `CourseDataAdapter` because it is being called by 3 fragments, that led to conflicts in defining the viewModel Type, If you want I can force cast them inside the adapter accordingly. Let me know your thoughts on this. And let me update the elvis operator with toString() wherever possible

aggarwalpulkit596 commented:
Why there is a courseDao Object in CourseDataAdapter it wasn't there before

adityastic commented on app/src/main/java/com/codingblocks/cbonlineapp/adapters/CourseDataAdapter.kt:
```diff
@@ -47,107 +39,18 @@ class CourseDataAdapter(
-        fun bindView(courseRun: CourseRun, courseWithInstructorDao: CourseWithInstructorDao, context: Context) {
-
-            database = AppDatabase.getInstance(context)
-            courseDao = database.courseDao()
```
This is why @aggarwalpulkit596

adityastic commented:
@aggarwalpulkit596 I've Claimed 1500 points for this as well. Please accept the claim in your free time
Regards 😄

adityastic commented:
@aggarwalpulkit596 bounty allotted to me for this pr is 1250 and not 1500. can you please check, I remember claiming 1500 only

aggarwalpulkit596 commented:
Yes i deducted 250 points for some reason i don't remember exactly why

adityastic commented:
> Yes i deducted 250 points for some reason i don't remember exactly why

Is there any way I can try to make up for them?

aggarwalpulkit596 commented:
Well, Now you can't as you can't raise a claim on the website anymore

adityastic commented:
> Well, Now you can't as you can't raise a claim on the website anymore

anything other than making a claim, can you not change it and let me fix it here?
